setUpdateRequestHeadersOverride throws InvalidRequestHeadersOverrideException when expo-channel-name is not pre-declared in app.config requestHeaders (local builds)

## Minimal reproducible example

https://gist.github.com/ihor-sviziev/c4d3e087b8555acc0710553692ce6511

**Steps:**

1. Create an app with `npx create-expo-app@latest --template blank-typescript`
2. Install `expo-updates`
3. Set `app.config.ts` with **no `requestHeaders`** under `updates`:
   ```ts
   updates: {
     url: 'https://u.expo.dev/<your-project-id>',
     // no requestHeaders
   }
   ```
4. Add a button that calls `Updates.setUpdateRequestHeadersOverride({ 'expo-channel-name': 'preview' })`
5. Build locally: `npx expo run:ios --configuration Release`
6. Tap the button

**Expected:** channel override is applied

**Actual:** crash on iOS:
```
FunctionCallException: Calling the 'setUpdateRequestHeadersOverride' function has failed
→ Caused by: InvalidRequestHeadersOverrideException: Invalid update requestHeaders override:
  Optional(["expo-channel-name": "preview"]) (at EXUpdates/EnabledAppController.swift:271)
```

> Note: this does **not** reproduce with EAS cloud builds because EAS automatically injects `requestHeaders` into the native binary when a channel is configured in `eas.json`. Only local builds are affected.

## Steps to reproduce

1. Clone the gist above
2. Run `npx create-expo-app@latest repro --template blank-typescript && cd repro`
3. Install expo-updates: `npx expo install expo-updates`
4. Replace `app.config.ts` with the one from the gist (no `requestHeaders`)
5. Add to `App.tsx`: `import * as Updates from 'expo-updates'` and a button calling `Updates.setUpdateRequestHeadersOverride({ 'expo-channel-name': 'preview' })`
6. Build: `npx expo run:ios --configuration Release`
7. Tap the button → error thrown

## Environment

```
expo-env-info 2.0.12 environment info:
    System:
      OS: macOS 26.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 24.15.0
      Yarn: 1.22.22
      npm: 11.12.1
      Watchman: 2026.04.20.00
    Managers:
      CocoaPods: 1.16.2
    SDKs:
      iOS SDK:
        Platforms: iOS 26.4, macOS 26.4
    IDEs:
      Xcode: 26.4.1/17E202
    npmPackages:
      expo: ~54.0.30 => 54.0.33
      expo-updates: ~29.0.16 => 29.0.16
      react: 19.1.0
      react-native: 0.81.5
    Expo Workflow: bare
```

## Expo Doctor Diagnostics

```
17/17 checks passed. No issues detected!
```

## Root Cause

`EnabledAppController.swift` calls `UpdatesConfig.isValidRequestHeadersOverride`, which requires every override key to exist in `originalEmbeddedRequestHeaders` — the headers baked into the binary from `app.config`:

```swift
// UpdatesConfig.swift
return overrideKeys.allSatisfy { !disallowHeaderKeys.contains($0) } &&
  overrideKeys.allSatisfy { originalEmbeddedKeys.contains($0) }
```

If `requestHeaders` is absent from `app.config`, `originalEmbeddedRequestHeaders` is `{}`, so any override key is rejected — including `expo-channel-name`.

## What the Docs Show

The [override documentation](https://docs.expo.dev/eas-update/override/) and [channel surfing blog post](https://expo.dev/blog/channel-surfing-for-expo-updates-how-to-switch-update-channels-at-runtime) both show the API with no prerequisite:

```js
Updates.setUpdateRequestHeadersOverride({ 'expo-channel-name': 'preview' });
```

No mention of `requestHeaders` being required in `app.config`.

## Workaround

Add `requestHeaders` to `app.config` before building:

```ts
updates: {
  url: 'https://u.expo.dev/<project-id>',
  requestHeaders: {
    'expo-channel-name': 'production', // or 'preview' for preview builds
  },
},
```

## Suggested Fix

Any of the following would resolve this:

**1. Document it** — add a note to the API reference and override guide that `expo-channel-name` must be present in `updates.requestHeaders` in `app.config` for local builds.

**2. Better error message** — instead of the generic exception, say something like:
> "The key 'expo-channel-name' is not in the embedded request headers. Add it to `updates.requestHeaders` in your app config and rebuild."

**3. Whitelist `expo-channel-name` implicitly** — since `setUpdateRequestHeadersOverride` was introduced specifically for channel surfing, `expo-channel-name` could be allowed without pre-declaration:

```swift
let alwaysAllowedKeys: Set<String> = ["expo-channel-name"]

return overrideKeys.allSatisfy { !disallowHeaderKeys.contains($0) } &&
  overrideKeys.allSatisfy { alwaysAllowedKeys.contains($0) || originalEmbeddedKeys.contains($0) }
```

*Security note*: The existing validation blocks overriding arbitrary headers like `Host` (request hijacking). Whitelisting only `expo-channel-name` preserves that protection for all other headers while making the primary use case of this API work without a build prerequisite. If there are security concerns specific to `expo-channel-name`, they apply equally to builds that already declare it — so the current behaviour provides no additional protection over option 3.

--- a/docs/pages/eas-update/override.mdx
+++ b/docs/pages/eas-update/override.mdx
@@ -51,7 +51,7 @@ await Updates.fetchUpdateAsync();
 await Updates.reloadAsync();
 ```
 
-> **info** Any header you want to override at runtime must be declared in `updates.requestHeaders` in your app config at build time.
+> **info** Any header you want to override at runtime must be declared in `updates.requestHeaders` in your app config.
 
 ### Risks and considerations when switching channels
 

--- a/docs/pages/eas-update/override.mdx
+++ b/docs/pages/eas-update/override.mdx
@@ -51,6 +51,8 @@ await Updates.fetchUpdateAsync();
 await Updates.reloadAsync();
 ```
 
+> **info** Any header you want to override at runtime must be declared in `updates.requestHeaders` in your app config at build time.
+
 ### Risks and considerations when switching channels
 
 Switching channels changes the JavaScript bundle the app runs. If your app depends on migrations or data shapes that are not compatible across channels, switching back and forth may cause issues.

--- a/packages/expo-updates/CHANGELOG.md
+++ b/packages/expo-updates/CHANGELOG.md
@@ -17,7 +17,7 @@
 
 - Pass absolute path to CLI helpers when creating build manifest, since the underlying functions now handle entry file inputs properly, instead of applying `mainModuleName` semantics to them ([#44414](https://github.com/expo/expo/pull/44414) by [@kitten](https://github.com/kitten))
 - [ios] Fix loading assets in brownfield ([#44724](https://github.com/expo/expo/pull/44724) by [@gabrieldonadel](https://github.com/gabrieldonadel))
-- Improve the error message thrown by `setUpdateRequestHeadersOverride` when an override key is not in `updates.requestHeaders`.  ([#45044](https://github.com/expo/expo/pull/45044) by [@alanjhughes](https://github.com/alanjhughes)) 
+- Improve the error message thrown by `setUpdateRequestHeadersOverride` when an override key is not in `updates.requestHeaders`. ([#45044](https://github.com/expo/expo/pull/45044) by [@alanjhughes](https://github.com/alanjhughes)) 
 
 ### 💡 Others
 

--- a/packages/expo-updates/CHANGELOG.md
+++ b/packages/expo-updates/CHANGELOG.md
@@ -17,6 +17,7 @@
 
 - Pass absolute path to CLI helpers when creating build manifest, since the underlying functions now handle entry file inputs properly, instead of applying `mainModuleName` semantics to them ([#44414](https://github.com/expo/expo/pull/44414) by [@kitten](https://github.com/kitten))
 - [ios] Fix loading assets in brownfield ([#44724](https://github.com/expo/expo/pull/44724) by [@gabrieldonadel](https://github.com/gabrieldonadel))
+- Improve the error message thrown by `setUpdateRequestHeadersOverride` when an override key is not in `updates.requestHeaders`. 
 
 ### 💡 Others
 

--- a/packages/expo-updates/CHANGELOG.md
+++ b/packages/expo-updates/CHANGELOG.md
@@ -17,7 +17,7 @@
 
 - Pass absolute path to CLI helpers when creating build manifest, since the underlying functions now handle entry file inputs properly, instead of applying `mainModuleName` semantics to them ([#44414](https://github.com/expo/expo/pull/44414) by [@kitten](https://github.com/kitten))
 - [ios] Fix loading assets in brownfield ([#44724](https://github.com/expo/expo/pull/44724) by [@gabrieldonadel](https://github.com/gabrieldonadel))
-- Improve the error message thrown by `setUpdateRequestHeadersOverride` when an override key is not in `updates.requestHeaders`. 
+- Improve the error message thrown by `setUpdateRequestHeadersOverride` when an override key is not in `updates.requestHeaders`.  ([#45044](https://github.com/expo/expo/pull/45044) by [@alanjhughes](https://github.com/alanjhughes)) 
 
 ### 💡 Others
 

--- a/packages/expo-updates/android/src/main/java/expo/modules/updates/EnabledUpdatesController.kt
+++ b/packages/expo-updates/android/src/main/java/expo/modules/updates/EnabledUpdatesController.kt
@@ -323,7 +323,13 @@ class EnabledUpdatesController(
       requestHeaders
     )
     if (!isValidRequestHeaders) {
-      throw CodedException("ERR_UPDATES_RUNTIME_OVERRIDE", "Invalid update requestHeaders override: $requestHeaders", null)
+      throw CodedException(
+        "ERR_UPDATES_RUNTIME_OVERRIDE",
+        "Invalid update requestHeaders override: $requestHeaders. " +
+          "Override keys must be declared in `updates.requestHeaders` in your app config " +
+          "at build time. Add the key to `updates.requestHeaders` and rebuild the app.",
+        null
+      )
     }
     val configOverride = UpdatesConfigurationOverride.saveRequestHeaders(context, requestHeaders)
     updatesConfiguration = UpdatesConfiguration.create(context, updatesConfiguration, configOverride)

--- a/packages/expo-updates/ios/EXUpdates/Exceptions.swift
+++ b/packages/expo-updates/ios/EXUpdates/Exceptions.swift
@@ -91,7 +91,11 @@ internal final class InvalidRequestHeadersOverrideException: Exception {
   }
 
   override var reason: String {
-    "Invalid update requestHeaders override: \(String(describing: requestHeaders))"
+    """
+    Invalid update requestHeaders override: \(String(describing: requestHeaders)). \
+    Override keys must be declared in `updates.requestHeaders` in your app config at \
+    build time. Add the key to `updates.requestHeaders` and rebuild the app.
+    """
   }
 }
 

